.travis.yml

```diff
@@ -45,64 +50,119 @@ sudo: false
-      - git-core
-      - build-essential
-      - g++-5
-      - libboost-all-dev
-      - libhdf5-serial-dev
-      - libfftw3-dev
-      - python-dev
-      - python3-dev
-      - liblapack-dev
-      - libxml2-dev
-      - libxslt-dev
-      - libarmadillo-dev
-      - libace-dev
-      - libgtest-dev
-      - libplplot-dev
-      - root-system-bin
-      - swig
+ apt:
+  sources:
+   - ubuntu-toolchain-r-test
+  packages:
+   - git-core
+   - build-essential
+   - g++-5
+   - libboost-all-dev
+   - libhdf5-serial-dev
+   - libfftw3-dev
+   - python-dev
+   - python3-dev
+   - python-tk
+   - python3-tk
+   - liblapack-dev
+   - libxml2-dev
+   - libarmadillo-dev
+   - libgtest-dev
+   - libplplot-dev
+   - swig
+   - ccache
+   # not in whitelist
+   - libxslt-dev
+   - libace-dev
+   - root-system-bin
+
+# cache C/C++/pip
+cache:
+  - ccache
+  - pip
 
 before_install:
-  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew update; fi
-  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew tap homebrew/science; fi
-  - # Note: boost is already installed on osx on travis so should not be included. However, we currently need boost-python
-  - # if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew install boost-python; fi
-  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew install ace; fi
-  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew install swig; fi
-  - #if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew install scipy matplotlib; fi
-  - # need curl to get more recent cmake
-  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew install curl; fi
-  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then curl -0 https://cmake.org/files/v3.8/cmake-3.8.0-Darwin-x86_64.tar.gz -o cmake.tar.gz; fi
-  - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then curl -0 https://cmake.org/files/v3.7/cmake-3.7.2-Linux-x86_64.tar.gz -o cmake.tar.gz; fi
-  - tar xzf cmake.tar.gz
-  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then mv cmake-*/CMake.app/Contents/* cmake-*; fi
+ - mkdir -p ~/.local/bin
+ - pushd ~/.local/bin
+ - |
+    if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then
+      brew update
+      brew tap homebrew/science
+      # Note: boost is already installed on osx on travis
+      # so should not be included. However, we currently need boost-python
+      # brew install boost-python
+      brew install ace
+      brew install swig
+      brew install ccache
+      # brew install scipy matplotlib
+      # need curl to get more recent cmake
+      brew install curl
+      curl -0 https://cmake.org/files/v3.8/cmake-3.8.0-Darwin-x86_64.tar.gz -o cmake.tar.gz
+      tar xzf cmake.tar.gz
+      mv cmake-*/CMake.app/Contents/* cmake-*
+    elif [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
+      curl -0 https://cmake.org/files/v3.7/cmake-3.7.2-Linux-x86_64.tar.gz -o cmake.tar.gz
+      tar xzf cmake.tar.gz
+    fi
+ # cmake override
+ - pushd cmake-*/bin
+ - export PATH="$PWD:$PATH"
+ - popd
+ # get pip
+ - curl -0 https://bootstrap.pypa.io/get-pip.py -o get-pip.py
+ - python$PYMVER get-pip.py --user
```

Comment:
no guarantee that this python will be the one found by cmake. would have to enforce it in the cmake call by setting `PYTHON_EXECUTABLE`. For STIR, I did it the other way round: [first run CMake, then use whatever python it found](https://github.com/UCL/STIR/blob/f9a93a73dc0ee6ed7495fa92991a58264f9378c6/.travis.yml#L86).

Reply:
just tried (https://travis-ci.org/CCPPETMR/SIRF/builds/318679147), but it's going to fail since the libs and executable versions are still mismatched, don't want to look into getting them to match (2.7.10 bin, 2.7.14 lib) - anyone else have any ideas?